dynamic_programming/fibonacci_numbers.nim

```diff
@@ -0,0 +1,186 @@
+## Fibonacci Numbers
+## ===========================================================================
+## Fibonacci numbers are numbers that are part of the Fibonacci sequence.
+## The Fibonacci sequence generally starts with 0 and 1, every next number
+## is the sum of the two preceding numbers:
+##   - F(0) = 0;
+##   - F(1) = 1;
+##   - F(n) = F(n-1) + F(n-2);
+##
+##  References:
+##    - https://en.wikipedia.org/wiki/Fibonacci_sequence
+##    - https://oeis.org/A000045
+{.push raises: [].}
+
+runnableExamples:
+
+  doAssert fibonacciSeqIterative(5) == @[Natural(0), 1, 1, 2, 3]
+
+  doAssert fibonacciClosure(4) == Natural(3)
+
+  for f in fibonacciIterator(1.Natural..4.Natural):
+    echo f # 1, 1, 2, 3
+
+import std/math
+
+type
+  fibMatrix = array[2, array[2, int]]
+
+func `*`(m1, m2: fibMatrix): fibMatrix =
+  let
+    a = m1[0][0] * m2[0][0] + m1[0][1] * m2[1][0]
+    b = m1[0][0] * m2[0][1] + m1[0][1] * m2[1][1]
+    z = m1[1][0] * m2[0][0] + m1[1][1] * m2[1][0]
+    y = m1[1][0] * m2[0][1] + m1[1][1] * m2[1][1]
+
+  [[a, b], [z, y]]
+
+func pow(matrix: fibMatrix, n: Natural): fibMatrix =
+  ## binary matrix exponentiation (divide-and-conquer approach)
+  if n in {0, 1}:
+    matrix
```

Comment:
This looks suspicious to me. If $n == 0$, you want to return the identity matrix instead.

Reply:
Ok.

Reply:
How come this change had no impact on the tests?
I am surprised that tests passed before.

Reply:
I guess, because the only case where n == 0 in this algorithm is F(1) and the Initial Fibonacci matrix [0][0] coincidentally matches Identity Matrix [0][0].

Oh, and I also just skip 0 and 1, because they're trivial in fibonacci sequence, so F(1) never even calls the matrix implementation.
```Nim
  if nth <= 1: return nth
```

Reply:
This doesn't explain why the result of pow was correct.
All calls to pow eventually call pow(0) or pow(1).

Why are you performing binary exponentiation recursively?
You are adding all intermediate matrices to the call stack.
You can look at: https://github.com/jilljenn/tryalgo/blob/master/tryalgo/fast_exponentiation.py
for two iterative versions of the algorithm.
Remove the asserts, the modulo ops., make it Nimesque and apply it to your matrix case.
Notice that if $n$ is odd, then $n-1$ is necessary even, so you can combine in this case the two operations to reduce the number of recursive calls.

Reply:
Thank you for the feedback and the link.
Honestly, I had to grab a pen and paper to walk through the algorithm step by step, but now, I think, your last sentence is much clearer to me.

I've added the second algorithm for iterative pow from your link and also a couple tests for the exponentiation function.